File: app/src/main/java/com/abishov/hexocat/common/picasso/CircleTransformation.java
package com.abishov.hexocat.common.picasso;

import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;

import com.squareup.picasso.Transformation;

import javax.inject.Inject;
import javax.inject.Singleton;

import static android.graphics.Bitmap.Config.ARGB_8888;
import static android.graphics.Paint.ANTI_ALIAS_FLAG;
import static android.graphics.Shader.TileMode.CLAMP;

@Singleton
final class CircleTransformation implements Transformation {

    @Inject
    CircleTransformation() {
        // keeping this constructor for dagger
    }

    @Override
    public Bitmap transform(Bitmap bitmap) {
        int size = bitmap.getWidth();
        Bitmap rounded = Bitmap.createBitmap(size, size, ARGB_8888);
        Canvas canvas = new Canvas(rounded);

        BitmapShader shader = new BitmapShader(bitmap, CLAMP, CLAMP);
        Paint shaderPaint = new Paint(ANTI_ALIAS_FLAG);
        shaderPaint.setShader(shader);

        RectF rect = new RectF(0, 0, size, size);
        float radius = size / 2f;
        canvas.drawRoundRect(rect, radius, radius, shaderPaint);

        bitmap.recycle();
        return rounded;
    }

    @Override
    public String key() {
        return "circle";
    }
}

File: app/src/main/java/com/abishov/hexocat/common/picasso/CircleTransformation.kt
package com.abishov.hexocat.common.picasso

import android.graphics.*
import android.graphics.Bitmap.Config.ARGB_8888
import android.graphics.Paint.ANTI_ALIAS_FLAG
import android.graphics.Shader.TileMode.CLAMP
import com.squareup.picasso.Transformation
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
internal class CircleTransformation @Inject constructor() : Transformation {

  override fun transform(bitmap: Bitmap): Bitmap {
    val size = bitmap.width
    val rounded = Bitmap.createBitmap(size, size, ARGB_8888)
    val canvas = Canvas(rounded)

    val shader = BitmapShader(bitmap, CLAMP, CLAMP)
    val shaderPaint = Paint(ANTI_ALIAS_FLAG)
    shaderPaint.shader = shader

    val rect = RectF(0f, 0f, size.toFloat(), size.toFloat())
    val radius = size / 2f
    canvas.drawRoundRect(rect, radius, radius, shaderPaint)

    bitmap.recycle()
    return rounded
  }

  override fun key() = "circle"
}